core/src/main/scala/kafka/server/AuthHelper.scala

```diff
@@ -130,4 +134,57 @@ class AuthHelper(authorizer: Option[Authorizer]) {
     }
   }
 
+  def computeDescribeClusterResponse(
+    request: RequestChannel.Request,
+    expectedEndpointType: EndpointType,
+    clusterId: String,
+    getNodes: () => DescribeClusterBrokerCollection,
+    getControllerId: () => Int
+  ): DescribeClusterResponseData = {
+    val describeClusterRequest = request.body[DescribeClusterRequest]
+    val requestEndpointType = EndpointType.fromId(describeClusterRequest.data().endpointType())
+    if (requestEndpointType.equals(EndpointType.UNKNOWN)) {
+      return new DescribeClusterResponseData().
+        setErrorCode(if (request.header.data().requestApiVersion() == 0) {
+          Errors.INVALID_REQUEST.code()
+        } else {
+          Errors.UNSUPPORTED_ENDPOINT_TYPE.code()
+        }).
+        setErrorMessage("Unsupported endpoint type " + describeClusterRequest.data().endpointType().toInt)
+    } else if (!expectedEndpointType.equals(requestEndpointType)) {
+      return new DescribeClusterResponseData().
+        setErrorCode(if (request.header.data().requestApiVersion() == 0) {
+          Errors.INVALID_REQUEST.code()
+        } else {
+          Errors.MISMATCHED_ENDPOINT_TYPE.code()
+        }).
+        setErrorMessage("The request was sent to an endpoint of type " + expectedEndpointType +
+          ", but we wanted an endpoint of type " + requestEndpointType)
+    }
```

Comment:
Looks like we won't return `INVALID_REQUEST` response when `request API version == 0`, and all other info are valid (i.e. endpoint type is correct and matching). Is that expected? I think we should throw `INVALID_REQUEST` error in any case?

Reply:
It's not possible to have request API version 0 and have an endpoint type other than BROKER. That's enforced by the serialization layer, so we don't have to do anything here.

Reply:
Yes, but that's under the assumption of users are using our java client. What if they use other client that no validation during serialization?

Reply:
v0 of the schema has no place to put endpoint type. it simply cannot be serialized in the v0 encoding. doesn't matter what client or programming language you use.

Reply:
OK, thanks.